crates/cacache/src/content/read.rs

```diff
@@ -1,100 +1,164 @@
+        let cpath = path::content_path(&cache, &sri);
+        let sri = sri.clone();
+        let mut reader = Self::instantiate(cpath, sri)?;
+
+        let mut v = Vec::with_capacity(reader.expected_size);
+        reader.read_to_end(&mut v).to_internal()?;
+        reader.check()?;
+        Ok(v)
+    }
+
+    pub async fn consume_async(cache: &Path, sri: &Integrity) -> Result<Vec<u8>> {
+        let cpath = path::content_path(&cache, &sri);
+        let sri = sri.clone();
+        async_std::task::spawn_blocking(|| {
+            let mut reader = Self::instantiate(cpath, sri)?;
+
+            let mut v = Vec::with_capacity(reader.expected_size);
+            reader.read_to_end(&mut v).to_internal()?;
+            reader.check()?;
+            Ok(v)
+        })
+        .await
     }
 }
 
 pub fn open(cache: &Path, sri: Integrity) -> Result<Reader> {
-    let cpath = path::content_path(&cache, &sri);
-    Ok(Reader {
-        fd: File::open(cpath).to_internal()?,
-        checker: IntegrityChecker::new(sri),
-    })
+    Reader::new(cache, &sri)
 }
 
-pub async fn open_async(cache: &Path, sri: Integrity) -> Result<AsyncReader> {
-    let cpath = path::content_path(&cache, &sri);
-    Ok(AsyncReader {
-        fd: async_std::fs::File::open(cpath).await.to_internal()?,
-        checker: IntegrityChecker::new(sri),
-    })
+pub async fn open_async(cache: &Path, sri: Integrity) -> Result<Reader> {
+    Reader::new_async(cache, &sri).await
 }
 
 pub fn read(cache: &Path, sri: &Integrity) -> Result<Vec<u8>> {
-    let cpath = path::content_path(&cache, &sri);
-    let ret = fs::read(&cpath).to_internal()?;
-    sri.check(&ret)?;
-    Ok(ret)
+    Reader::consume(cache, sri)
 }
 
-pub async fn read_async<'a>(cache: &'a Path, sri: &'a Integrity) -> Result<Vec<u8>> {
-    let cpath = path::content_path(&cache, &sri);
-    let ret = async_std::fs::read(&cpath).await.to_internal()?;
-    sri.check(&ret)?;
-    Ok(ret)
+pub async fn read_async<'a>(cache: &Path, sri: &Integrity) -> Result<Vec<u8>> {
+    Reader::consume_async(cache, sri).await
 }
 
 pub fn copy(cache: &Path, sri: &Integrity, to: &Path) -> Result<u64> {
-    let cpath = path::content_path(&cache, &sri);
-    let ret = fs::copy(&cpath, to).to_internal()?;
-    let data = fs::read(cpath).to_internal()?;
-    sri.check(data)?;
+    let mut reader = Reader::new(cache, sri)?;
+    let mut target = fs::OpenOptions::new()
+        .read(true)
+        .write(true)
+        .create(true)
+        .truncate(true)
+        .open(to)
+        .to_internal()?;
+
+    let ret = if reader.expected_size > 0 {
+        if let Ok(mut mmap) = unsafe { MmapMut::map_mut(&target) } {
+            let mut cursor = std::io::Cursor::new(&mut mmap[..]);
+            std::io::copy(&mut reader, &mut cursor).to_internal()
+        } else {
+            std::io::copy(&mut reader, &mut target).to_internal()
+        }
+    } else {
+        std::io::copy(&mut reader, &mut target).to_internal()
+    }?;
```

Comment:
This needs to be benchmarked & bounded by min/max eligible sizes. I'd like to follow up with a PR to add more benchmarks / CI tooling around benchmarks.